pydm/display.py

```diff
@@ -83,16 +93,24 @@ def _compile_ui_file(uifile: str) -> Tuple[str, str]:
     ----------
     uifile : str
         The path to a .ui file to compile
+    macro_keys : FrozenSet, optional
+        Only the keys of any macros associated with the display. FrozenSet so that it hashes for the lru_cache
 
     Returns
     -------
     Tuple[str, str] - The first element is the compiled ui file, the second is the name of the class (e.g. Ui_Form)
     """
     code_string = StringIO()
     uic.compileUi(uifile, code_string)
+    compiled_code = code_string.getvalue()
+    if macro_keys is not None and len(macro_keys) > 0:
+        # Replaces any single quoted macro definitions with double quoted ones
```

Comment:
Could this approach cause a problem if the user passes in a macro that has double quotes, instead of single quotes as in the original bug example?
e.g. I assume this user set their macro to `xterm -e 'command to run'` but what if instead they had picked `xterm -e "command to run"`?
```
self.PyDMShellCommand.setProperty("commands", ["xterm -e "command to run""])
```
Maybe it could be worth it to go all the way to triple quotes if this is a problem?
```
self.PyDMShellCommand.setProperty("commands", ["""xterm -e "command to run""""])
self.PyDMShellCommand.setProperty("commands", ["""xterm -e 'command to run'"""])
```
Disregard if I've misunderstood the issue

Reply:
(I also feel like I don't fully grasp the issue so - grain of salt)
Alternatively could escaping all single quotes in all macro values prior to interpolating the code work? i.e., `"'" -> "\\'"`

Reply:
Oh, that would be much more straightforward than what I am doing haha and should work fine. The original issue is that in a json file for macro replacement for a template repeater are lines like this:

`"button_1_command":"xterm -e 'echo hello'"`

which end up compiled by uic to:

`self.PyDMShellCommand.setProperty("commands", ['${button_1_command}'])`

and finally the macro replacement results in invalid python code:

`self.PyDMShellCommand.setProperty("commands", ['xterm -e 'echo hello''])`

So escaping the quotes before doing the macro replacement should work fine. I can do it via regex to ignore already escaped single quotes just in case someone is already has `\'` in their macros.